src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+                                    if target resolution is not specified"""
+                                )
+
+                if full_extents:
+                    full_extents = BoundingBox(
+                        *shapely.unary_union(
+                            [
+                                shapely.box(*full_extents.bounds),
+                                shapely.box(*extents.bounds),
+                            ]
+                        ).bounds
+                    )
+                else:
+                    full_extents = extents
+
+                meta.append(
+                    GeoMetadata(
+                        paths=pth,
+                        extents=extents,
+                        crs=crs,
+                    )
+                )
+            if len(meta) > 0:
+                blocks = group_by_raster_block(meta)
+                return GeoMetadata(
+                    block_metadata=blocks,
+                    crs=crs,
+                    extents=full_extents,
+                    res=(xres, yres),
+                    band_count=num_bands,
+                    dtype=dtype,
+                )
+            else:
+                raise ValueError("Raster datasets not found")
+
+
+def ingest_geometry_udf(
+    sources: Optional[Sequence[str]] = None,
+    *,
+    dataset_uri: str,
+    schema: dict,
+    extents: Optional[Tuple[float, float, float, float]] = None,
+    crs: Optional[str] = None,
+    chunk_size: Optional[int] = GEOMETRY_CHUNK_SIZE,
+    compressor: Optional[dict] = None,
+    append: bool = False,
+    verbose: bool = False,
+    config: Optional[Mapping[str, Any]] = None,
+    id: str = "geometry",
+    trace: bool = False,
+    log_uri: Optional[str] = None,
+):
+    """Internal udf that ingests server side batch of geometry files
+    into tiledb arrays using Fiona API
+
+    :param sources: Sequence of input geometry file names
+    :param dataset_uri: str, output TileDB array name
+    :param schema: dict, dictionary of schema attributes and geometries
+    :param geometry_type: str, fiona geometry type e.g. `Point`
+    :param extents: Extents of the destination geometry array
+    :param crs: str, CRS for the destination dataset
+    :param chunk_size: int, sets tile capacity and
+                       the number of geometries written at once
+    :param compressor: dict, serialized compression filter
+    :param append: bool, whether to append to the array
+    :param verbose: verbose logging, defaults to False
+    :param config: dict, configuration to pass on tiledb.VFS
+    :param id: str, ID for logging, defaults to None
+    :param trace, bool, enable trace logging
+    :param log_uri: log array URI
+    """
+    import fiona
+
+    from tiledb import filter
+
+    with tiledb.scope_ctx(config):
+        with Profiler(array_uri=log_uri, id=id, trace=trace):
+            tiledb.VFS()
+            logger = get_logger_wrapper(verbose)
+
+            if append:
```

Comment:
Looking at how all these if-blocks are run, it seems like in the general case only one of them is ever going to run and they're the last thing in the function (except the logging statement). It might make more sense to structure it with early returns. To maintain the logging, we can use a try-finally:

```python
try:
  if append:
     ...
     return
  if extends and crs:
    ...
    return
finally:
  logger.info("max memory usage: %.3f GiB", max_memory_usage() / (1 << 30))
```

Reply:
👍